primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -75,10 +77,133 @@ where
 
     /// Enforce correct `merkle_proof` for the `elem` against
     /// `expected_merkle_root`.
-    fn enforce_merkle_proof(
+    fn enforce_membership_proof(
         &mut self,
         elem: Self::LeafVar,
         merkle_proof: Self::MerklePathVar,
         expected_merkle_root: Variable,
     ) -> Result<(), CircuitError>;
 }
+
+/// Gadget for the sparse merkle tree
+pub trait SparseMerkleTreeGadget<M>: MerkleTreeGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    /// checking non-membership proof
+    fn is_non_member(
+        &mut self,
+        elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        merkle_root: Variable,
+    ) -> Result<BoolVar, CircuitError>;
+
+    /// Enforce correct `merkle_proof` for the empty leaf `elem` against
+    /// `expected_merkle_root`.
+    fn enforce_non_membership_proof(
+        &mut self,
+        empty_elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        expected_merkle_root: Variable,
+    ) -> Result<(), CircuitError>;
+}
+
+pub(crate) trait MerkleTreeHelperGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    type MembershipProofBooleanEncoding;
+    type MembershipProofVar;
```

Comment:
add some documentation on these types? I have some trouble understanding (the semantic meaning of) `BooleanEncoding`, also the mixed usage of `merkle_path`, `merkle_proof` and `memebership_proof`.

Reply:
Right, the path/proof terminology should definitely be reworked.

Reply:
Here's my attempt at separating the path/proof, as well as better name for encoding. https://github.com/EspressoSystems/jellyfish/pull/172/commits/767d8d9c0fd9743add2dbc9c3d4a8cc277bf93ac
@EspressoSystems/jellyfish Would be good if we all reach consensus on the naming.